### PR TITLE
fix: escape shell arg before passing into curl body

### PR DESCRIPTION
**Pull request by oakbani**
## Summary
User ID gets feeded into curl body unsanitized. This is a quick fix to prevent any arbitrary command being executed. [escapeshellarg](https://www.php.net/manual/en/function.escapeshellarg.php) escapes a string by surrounding it by single quotes and escaping any existing single quotes. This makes userId a safe shell argument. 

**_Concern_**
Since user ID will be surrounded by single quotes, the back end may consider it a different user ID if the user was activated or tracked using a different event dispatcher. 

## Test plan
Tested manually. All existing unit tests pass. 

## Issues
- OASIS-4426

### CODE REVIEW

**Comment by dj-mcculloch** [comment]
@aliabbasrizvi what if there are other evals being run on user input? I'd prefer to see user input be escaped/sanitized as early in the chain as possible to negate any future risk.

**Comment by aliabbasrizvi** [comment]
@dj-mcculloch I agree. I do not think there is other such evaluation elsewhere in the code however.
The concern mentioned in the summary is the reason I am suggesting this.

**Comment by aliabbasrizvi** [comment]
Also, perhaps we can use `escapeshellcmd`?

**Comment by oakbani** [comment]
@aliabbasrizvi  escapeshellarg works for single argument. I tried escapeshellcmd in CurlEventDispatcher body, but escapeshellcmd expects string, not an array. If I apply, escapeshellcmd on json encoded body, the escaping makes it an invalid curl request. 

I think only attribute value is the only other one susceptible to similar behavior. Have escaped it when it's a string. Let me know if this works.

File: src/Optimizely/Event/Builder/EventBuilder.php [hunk 1]
```diff
@@ -106,6 +106,8 @@ private function getCommonParams($config, $userId, $attributes)
                 if (Validator::isAttributeValid($attributeKey, $attributeValue)) {
                     $attributeEntity = $config->getAttribute($attributeKey);
                     if ($attributeEntity instanceof Attribute) {
+                        // escape single quotes if attribute value is of type string.
+                        $attributeValue = is_string($attributeValue) ? escapeshellarg($attributeValue): $attributeValue;
```
Comment (oakbani) [inline src/Optimizely/Event/Builder/EventBuilder.php]: applying escapeshellarg on any other type converts it into string.

**Comment by aliabbasrizvi** [comment]
@oakbani ok, I would like to however explore the possibility of containing this to `CurlEventDispatcher` only given the concern that you brought up.

**Comment by aliabbasrizvi** [comment]
Also, we have event tags which can contain string values?

**Comment by oakbani** [comment]
on another thought I think this is the right place to do this. Escaped
userIds and attribute values will remain consistent irrespective of the
dispatcher used.

File: src/Optimizely/Event/Dispatcher/CurlEventDispatcher.php [hunk 2]
```diff
@@ -34,12 +34,70 @@ public function dispatchEvent(LogEvent $event)
...
+     * 1. user ID.
+     * 2. attribute values of type string.
+     * 3. event tag keys.
+     * 4. event tag values of type string.
+     *
+     * @param  array  $params LogEvent params.
+     *
+     * @return array Escaped params.
+     **/
+    public function sanitizeEventPayload(array $params)
```
Comment (aliabbasrizvi) [inline src/Optimizely/Event/Dispatcher/CurlEventDispatcher.php]: nit. Minor feedback about code arrangement. I feel this method should be declared earlier and then referenced here.

File: src/Optimizely/Event/Dispatcher/CurlEventDispatcher.php [hunk 3]
```diff
@@ -27,14 +27,74 @@
...
+                $attr['value'] = escapeIfString($attr['value']);
+                 array_push($escapedAttributes, $attr);
+            }
+
+            $visitor['attributes'] = $escapedAttributes;
+
+            // event tags if present
+            if (isset($visitor['snapshots'][0]['events'][0]['tags'])) {
+                $eventTags = $visitor['snapshots'][0]['events'][0]['tags'];
+                $escapedEventTags = [];
```
Comment (aliabbasrizvi) [inline src/Optimizely/Event/Dispatcher/CurlEventDispatcher.php]: Needs to be declared outside of the if condition.

File: src/Optimizely/Event/Dispatcher/CurlEventDispatcher.php [hunk 4]
```diff
@@ -27,14 +27,74 @@
...
+     * 4. event tag values of type string.
+     *
+     * @param  array  $params LogEvent params.
+     *
+     * @return array Escaped params.
+     **/
+    public function sanitizeEventPayload(array $params)
+    {
+
+        function escapeIfString($val)
```
Comment (aliabbasrizvi) [inline src/Optimizely/Event/Dispatcher/CurlEventDispatcher.php]: This should be a class level method probably as on every call to `sanitizeEventPayload` this triggers a redeclare method error. Please fix.

**Pull request merged**